omnibus/config/software/libjemalloc.rb

```diff
@@ -0,0 +1,44 @@
+# Unless explicitly stated otherwise all files in this repository are licensed
+# under the Apache License Version 2.0.
+# This product includes software developed at Datadog (https:#www.datadoghq.com/).
+# Copyright 2022-present Datadog, Inc.
+
+name "libjemalloc"
+default_version "5.3.0"
+
+version "5.3.0" do
+  source sha256: "ef6f74fd45e95ee4ef7f9e19ebe5b075ca6b7fbe0140612b2a161abafb7ee179" 
+end
+
+ship_source_offer true
+
+source url:"https://github.com/jemalloc/jemalloc/archive/refs/tags/#{version}.tar.gz",
+       extract: :seven_zip
+
+relative_path "jemalloc-#{version}"
+
+build do
+    license "BSD-2-Clause"
+    license_file "./COPYING"
+
+    env = with_standard_compiler_flags
+
+    # By default jemalloc releases pages using madvise with MADV_FREE which tells the
+    # OS to reclaim pages lazily.
+    # This can result in misleading metrics, so we need to disable it here.
+    # https://github.com/jemalloc/jemalloc/blob/dev/INSTALL.md#advanced-configuration
+    env["je_cv_madv_free"] = "no"
```

Comment:
Can we test that this has the intended effect? I'd just like to double check since I'm not familiar enough with autoconf / configure here to confirm whether this is what @chouquette meant.

Reply:
AFAICS from jemalloc's build system the only way to ensure this has an effect is to check the content of `include/jemalloc/internal/jemalloc_internal_defs.h` and ensures it has `JEMALLOC_PURGE_MADVISE_FREE` defined.

The quick and dirty way would be to add a grep in the build recipe I suppose?

Reply:
I have quite heavily benchmarked the agent with and without this setting and not found it to make a difference for us. This is largely here because a number of other projects using `jemalloc` have had problems and found it necessary to set this. So it is largely here 'just in case'.

To verify that `configure` is picking up the setting, we should be able to see from the output of `./configure` this line:

```
checking whether madvise(..., MADV_FREE) is compilable... (cached) no
```
Looking at the job [here](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/655716818#L2317), it doesn't look like the output is being captured. Do you know if there is somewhere else we could look? Do we have the `config.log` around anywhere?

Reply:
> The quick and dirty way would be to add a grep in the build recipe I suppose?

Ah, I missed this suggestion. I'll have a go.

Reply:
```
diff --git a/.gitlab/package_build/linux.yml b/.gitlab/package_build/linux.yml
index 3b3d08ac8c..58aa4ead82 100644
--- a/.gitlab/package_build/linux.yml
+++ b/.gitlab/package_build/linux.yml
@@ -12,7 +12,7 @@
   - $S3_CP_CMD $S3_PERMANENT_ARTIFACTS_URI/llc-$CLANG_LLVM_VER.${PACKAGE_ARCH} /tmp/system-probe/llc-bpf
   - cp $CI_PROJECT_DIR/minimized-btfs.tar.xz /tmp/system-probe/minimized-btfs.tar.xz
   - chmod 0744 /tmp/system-probe/clang-bpf /tmp/system-probe/llc-bpf
-  - inv -e omnibus.build --release-version "$RELEASE_VERSION" --major-version "$AGENT_MAJOR_VERSION" --python-runtimes "$PYTHON_RUNTIMES" --base-dir $OMNIBUS_BASE_DIR  ${USE_S3_CACHING} --skip-deps --go-mod-cache="$GOPATH/pkg/mod" --system-probe-bin=/tmp/system-probe --flavor "$FLAVOR" --config-directory "$CONFIG_DIR" --install-directory "$INSTALL_DIR"
+  - inv -e omnibus.build --release-version "$RELEASE_VERSION" --major-version "$AGENT_MAJOR_VERSION" --python-runtimes "$PYTHON_RUNTIMES" --base-dir $OMNIBUS_BASE_DIR  ${USE_S3_CACHING} --skip-deps --go-mod-cache="$GOPATH/pkg/mod" --system-probe-bin=/tmp/system-probe --flavor "$FLAVOR" --config-directory "$CONFIG_DIR" --install-directory "$INSTALL_DIR" --log-level debug
   - ls -la $OMNIBUS_PACKAGE_DIR
   - !reference [.upload_sbom_artifacts]
```
will capture the output of `configure`

Indeed I missed that they run a specific check when `je_cv_madvise` is passed so that will also work!

Reply:
`  [Software: libjemalloc] I | 2024-10-02T10:42:02+00:00 | Restored from cache`
seems like our cache ignores the build command line, which sucks. I'll write that down in our ongoing gathering of build system pain points.
In the meantime, you can add a dummy environment variable to the build, that should be enough to invalidate the cache 
as long as it isn't explicitly ignored by the cache config
something like `JEMALLOC_FORCE_BUILD: 1` will do the trick

Reply:
We have in [the job](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/658533264/raw):

```
   D | 2024-10-02T13:37:31+00:00 | checking whether madvise(..., MADV_FREE) is compilable... (cached) no
```

Reply:
I think I understand their configure script a bit better, and it's my mistake, setting the value as an environment variable will only work when `MADV_FREE` is detected as not available.
When it's available, `je_cv_madv_free` will be overridden to `yes` which will defeat the puspose.

The only way to prevent this check is to do as their documentation suggests, including passing `-C` to `./configure` to enable reading from `config.cache` which is disabled by default

Reply:
I confess I am a little confused. I get the same output if I try with an environment variable or by setting `config.cache`. But *both these outputs seem to be wrong*. In both cases `JEMALLOC_PURGE_MADVISE_FREE` is defined.

I note that CockroachDB have also fixed this using the environment variable, see [here](https://github.com/cockroachdb/cockroach/pull/83795/files).

I will need to dig into it a little further tomorrow...

Reply:
As far as I understand, in our case, the environment variable and/or `config.cache` is irrelevant:
```
   dnl Check for madvise(..., MADV_FREE).    
   JE_COMPILABLE([madvise(..., MADV_FREE)], [
 #include <sys/mman.h>                       
 ], [                                        
   madvise((void *)0, 0, MADV_FREE);       
 ], [je_cv_madv_free])                       
```
This will:
* check the cache for an already known value of `je_cv_madv_free`. If it is already known, the entire section is skipped. If not:
* attempt to compile a minimal code sample to figure out it `MADV_FREE` is available or not, and will set `je_cv_madv_free` if it is.

Judging from the `./configure` output, `MADV_FREE` isn't available, meaning `je_cv_madv_free` is either undefined or set to `no` from our environment variable.

However, if `MADV_FREE` was to become available (most likely from building on a machine with a more recent kernel), setting from the environment would then be overwritten since ` je_cv_madv_free` would be set by `JE_COMPILABLE`.

Now regarding `JEMALLOC_PURGE_MADVISE_FREE` being set, AFAICS this will always be the case if `madvise` is available & when building for linux x86_64:
```
if test "x${je_cv_madv_free}" = "xyes" ; then                
  AC_DEFINE([JEMALLOC_PURGE_MADVISE_FREE], [ ], [ ])         
elif test "x${je_cv_madvise}" = "xyes" ; then                
  case "${host_cpu}" in i686|x86_64)                         
     case "${host}" in *-*-linux*)                          
          AC_DEFINE([JEMALLOC_PURGE_MADVISE_FREE], [ ], [ ]) 
         AC_DEFINE([JEMALLOC_DEFINE_MADVISE_FREE], [ ], [ ])
       ;;                                                     
      esac                                                   
      ;;                                                     
  esac                                                       
```

Reply:
Ok, I see this is behaviour that has been introduced since version 5.0.1. I have updated to pull this version and will run some benchmarks against it.

Reply:
Ok. It turns out as per [this issue](https://github.com/jemalloc/jemalloc/issues/2728) `MADV_FREE` isn't actually an issue any more, so we don't need to do any of this at all!

Reply:
Well this simplifies things :grin: